kv/kvserver: TestBackpressureNotAppliedWhenReducingRangeSize failed

kv/kvserver.TestBackpressureNotAppliedWhenReducingRangeSize [failed](https://tanzanite.cluster.engflow.com/invocations/default/d3ccecba-14a9-4b61-aca2-d207b1d71b94?testReportRun=18&testReportShard=20&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [ecfb74b7ad75327f11814966a2cdab1b9793a549](https://github.com/cockroachdb/cockroach/commits/ecfb74b7ad75327f11814966a2cdab1b9793a549):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackpressureNotAppliedWhenReducingRangeSize (14m44s)
	TestBackpressureNotAppliedWhenReducingRangeSize/no_backpressure_when_much_larger_on_new_node (14m25s)

```
Stack: 

```
goroutine 19275 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* 
* goroutine 15151 [select]:
* github.com/cockroachdb/cockroach/pkg/util/retry.(*Retry).Next(0xc007b5fbf8)
* 	pkg/util/retry/retry.go:128 +0x1fd
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).getOrCreateReplica(0xc006de0008, {0x91cb610, 0xc00dcaa1e0}, 0x4b, 0x1, 0xc00b227a20)
* 	pkg/kv/kvserver/store_create_replica.go:73 +0x225
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).withReplicaForRequest(0xc006de0a48?, {0x91cb610, 0xc00dcaa1e0}, 0xc00b227a00, 0xc007b5fdd8)
* 	pkg/kv/kvserver/store_raft.go:405 +0x5a
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRequestQueue(0xc006de0008, {0x91cb610, 0xc00dcaa1e0}, 0x4b)
* 	pkg/kv/kvserver/store_raft.go:679 +0x1c9
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0013e6310, {0x91cb610, 0xc00dcaa1e0}, {0x91efa80, 0xc006de0008}, 0xc00bc48008)
* 	pkg/kv/kvserver/scheduler.go:397 +0x1c3
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x91cb610?, 0xc00dcaa1e0?})
* 	pkg/kv/kvserver/scheduler.go:319 +0x45
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x77d596e, 0xb})
* 	pkg/util/stop/stopper.go:500 +0x21d
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 15093
* 	pkg/util/stop/stopper.go:489 +0x47c
* 
* goroutine 15033 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0xbc
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 14979
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x58
* 
* goroutine 15087 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0xbc
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 14979
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x58
* 
* goroutine 15025 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	pkg/util/admission/work_queue.go:501 +0x14d
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 14979
* 	pkg/util/admission/work_queue.go:472 +0x4f
* 
* goroutine 19009 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc00d86e4e0, {0x91cb060, 0xc00d8a0dc0})
* 	GOROOT/src/database/sql/sql.go:1246 +0x87
* created by database/sql.OpenDB in goroutine 14979
* 	GOROOT/src/database/sql/sql.go:824 +0x14c
* 
* goroutine 18240 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 14979
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=18</code>
 - <code>shard=20</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackpressureNotAppliedWhenReducingRangeSize.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47534

--- a/pkg/kv/kvserver/raft_transport.go
+++ b/pkg/kv/kvserver/raft_transport.go
@@ -599,9 +599,9 @@ func (t *RaftTransport) StopOutgoingMessage(storeID roachpb.StoreID) {
 func (t *RaftTransport) processQueue(
 	ctx context.Context, q *raftSendQueue, client RPCMultiRaftClient, _ rpcbase.ConnectionClass,
 ) error {
-	batchCtx, cancel := context.WithCancel(ctx)
+	streamCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	stream, err := client.RaftMessageBatch(batchCtx) // closed via cancellation
+	stream, err := client.RaftMessageBatch(streamCtx) // closed via cancellation
 	if err != nil {
 		return errors.Wrapf(err, "creating batch client")
 	}
@@ -609,15 +609,14 @@ func (t *RaftTransport) processQueue(
 	errCh := make(chan error, 1)
 
 	// NB: the stream context is canceled when this func returns, and causes the
-	// response handling loop to terminate asynchronously.
-	//
-	// TODO(#140958): the context cancellation in the middle of HandleRaftResponse
-	// can lead to broken state, such as a replica marked as destroyed but this
-	// not being reflected in storage.
+	// response handling loop to terminate asynchronously. Note though that the
+	// task uses the parent context for HandleRaftResponse calls, to let it finish
+	// gracefully. Previously, context cancellation inside HandleRaftResponse
+	// could cause incorrect / half-done Replica destruction (see #140958).
 	//
 	// TODO(pav-kv): wait for the task termination to prevent subsequent
 	// processQueue calls from piling up concurrent tasks.
-	goCtx, hdl, err := t.stopper.GetHandle(stream.Context(), stop.TaskOpts{
+	goCtx, hdl, err := t.stopper.GetHandle(ctx, stop.TaskOpts{
 		TaskName: "storage.RaftTransport: processing queue",
 	})
 	if err != nil {

--- a/pkg/kv/kvserver/raft_transport.go
+++ b/pkg/kv/kvserver/raft_transport.go
@@ -597,13 +597,27 @@ func (t *RaftTransport) StopOutgoingMessage(storeID roachpb.StoreID) {
 // lost and a new instance of processQueue will be started by the next message
 // to be sent.
 func (t *RaftTransport) processQueue(
-	q *raftSendQueue, stream RPCMultiRaft_RaftMessageBatchClient, _ rpcbase.ConnectionClass,
+	ctx context.Context, q *raftSendQueue, client RPCMultiRaftClient, _ rpcbase.ConnectionClass,
 ) error {
+	batchCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	stream, err := client.RaftMessageBatch(batchCtx) // closed via cancellation
+	if err != nil {
+		return errors.Wrapf(err, "creating batch client")
+	}
+
 	errCh := make(chan error, 1)
 
-	ctx := stream.Context()
-
-	goCtx, hdl, err := t.stopper.GetHandle(ctx, stop.TaskOpts{
+	// NB: the stream context is canceled when this func returns, and causes the
+	// response handling loop to terminate asynchronously.
+	//
+	// TODO(#140958): the context cancellation in the middle of HandleRaftResponse
+	// can lead to broken state, such as a replica marked as destroyed but this
+	// not being reflected in storage.
+	//
+	// TODO(pav-kv): wait for the task termination to prevent subsequent
+	// processQueue calls from piling up concurrent tasks.
+	goCtx, hdl, err := t.stopper.GetHandle(stream.Context(), stop.TaskOpts{
 		TaskName: "storage.RaftTransport: processing queue",
 	})
 	if err != nil {
@@ -902,16 +916,7 @@ func (t *RaftTransport) startProcessNewQueue(
 			// DialNode already logs sufficiently, so just return.
 			return
 		}
-		batchCtx, cancel := context.WithCancel(ctx)
-		defer cancel()
-
-		stream, err := client.RaftMessageBatch(batchCtx) // closed via cancellation
-		if err != nil {
-			log.Warningf(ctx, "creating batch client for node %d failed: %+v", toNodeID, err)
-			return
-		}
-
-		if err := t.processQueue(q, stream, class); err != nil {
+		if err := t.processQueue(ctx, q, client, class); err != nil {
 			log.Warningf(ctx, "while processing outgoing Raft queue to node %d: %s:", toNodeID, err)
 		}
 	}

